tests/posix/common/src/pthread.c

```diff
@@ -465,6 +465,54 @@ ZTEST(pthread, test_pthread_cleanup)
 	zassert_ok(pthread_join(th, NULL));
 }
 
+static void *test_pthread_cancel_fn(void *arg)
+{
+	int oldtype = ~PTHREAD_CANCEL_DEFERRED;
+	int oldstate = ~PTHREAD_CANCEL_ENABLE;
+
+	zassert_ok(pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &oldtype));
+	zassert_equal(oldtype, PTHREAD_CANCEL_DEFERRED);
+
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &oldstate));
+	zassert_equal(oldstate, PTHREAD_CANCEL_ENABLE);
+
+	/* this should be ignored */
+	pthread_testcancel();
+
+	/* this will mark it pending */
+	pthread_cancel(pthread_self());
+
+	/* assume this completes well before 500ms */
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &oldstate));
+
+	/* this should terminate the thread*/
+	pthread_testcancel();
+
+	/*
+	 * We use this large delay to indicate that the thread in fact did not exit via
+	 * pthread_testcancel(), as expected. I.e. it should not take > 1000 ms to join thread
+	 */
+	zassert_ok(k_msleep(1000));
+
+	return NULL;
+}
+
+ZTEST(pthread, test_pthread_testcancel)
+{
+	uint32_t now;
+	uint32_t then;
+	pthread_t th;
+
+	then = k_uptime_get_32();
+	zassert_ok(pthread_create(&th, NULL, test_pthread_cancel_fn, NULL));
+	zassert_ok(k_msleep(500));
```

Comment:
we should make this much shorter, the test just need a delay to trigger a context switch for the `test_pthread_cancel_fn` thread to execute, which is expected to cancel itself almost immediately before it reaches the relatively longer sleep (`k_msleep(1000)`) at the end.